concent_api/core/payments/sci_callback.py

```diff
@@ -0,0 +1,149 @@
+        callback by SCI and return.
+    5b. Otherwise handle errors by raising an exception and interrupting the callback.
+
+    """
+    assert isinstance(transaction, Transaction)
+
+    # Create a TransactionSigningRequest.
+    transaction_signing_request = TransactionSigningRequest(
+        nonce    = transaction.nonce,
+        gasprice = transaction.gasprice,
+        startgas = transaction.startgas,
+        to       = transaction.to,
+        value    = transaction.value,
+        data     = transaction.data,
+    )
+
+    setattr(transaction_signing_request, 'from', settings.CONCENT_ETHEREUM_ADDRESS)
+
+    middleman_message = GolemMessageFrame(
+        payload=transaction_signing_request,
+        request_id=transaction.nonce,  # TODO: How to generate request_id ??
+    )
+
+    # Establish a TCP connection with Middleman.
+    with closing(socket.socket(socket.AF_INET, socket.SOCK_STREAM)) as client_socket:
+        client_socket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, True)
+        client_socket.connect((settings.MIDDLEMAN_ADDRESS, settings.MIDDLEMAN_PORT))
+        client_socket.settimeout(SCI_CALLBACK_MAXIMUM_TIMEOUT)
+
+        # Send the TransactionSigningRequest.
+        try:
+            send_over_stream(
+                connection=client_socket,
+                raw_message=middleman_message,
+                private_key=settings.CONCENT_PRIVATE_KEY,
+            )
+
+            # Read the response and close the connection.
+            raw_response = unescape_stream(connection=client_socket).__next__()
+        # Connection with Middleman times out before a complete response is received.
+        except socket.timeout as exception:
+            raise SCICallbackTimeoutError() from exception
+
+    try:
+        deserialized_message = GolemMessageFrame.deserialize(
+            raw_message=raw_response,
+            public_key=settings.SIGNING_SERVICE_PUBLIC_KEY,
+        )
+    # Received data frame is invalid or the signature does not match its content.
+    except MiddlemanProtocolError as exception:
+        raise SCICallbackFrameError() from exception
+    # If the received Golem message is malformed or not signed by the Signing Service.
+    except MessageError as exception:
+        raise SCICallbackPayloadError() from exception
+
+    # Request ID in the received data frame does not match the one in the data frame that was sent.
+    if deserialized_message.request_id != transaction.nonce:  # TODO: How to generate request_id ??
+        raise SCICallbackRequestIdError('MiddleMan response request_id does not match requested.')
+
+    # If the received Golem message is TransactionRejected or an ERROR frame.
+    if isinstance(deserialized_message, ErrorFrame) or isinstance(deserialized_message.payload, TransactionRejected):
+        raise SCICallbackPayloadError(
+            'Received frame is either ErrorFrame or contains Golem message TransactionRejected.'
```

Comment:
I think we should be more specific here

Reply:
I have split this into two conditions and added separate tests. :+1: